File: .github/scripts/ci_check.js
const fs = require("fs");
const path = require("path");

function setOutput(name, value) {
  if (process.env.GITHUB_OUTPUT) {
    fs.appendFileSync(process.env.GITHUB_OUTPUT, `${name}=${value}\n`);
  }
}

function parseCommitConfig(config) {
  if (!Array.isArray(config.commit) || config.commit.length < 2) {
    throw new Error(
      'ci.config.json "commit" must be an array like ["on", "main"].'
    );
  }

  const [toggleRaw, branchRaw] = config.commit;
  const toggle = String(toggleRaw).toLowerCase();
  const branch = String(branchRaw || "").trim();

  if (!["on", "off"].includes(toggle)) {
    throw new Error('ci.config.json commit toggle must be "on" or "off".');
  }

  if (!branch) {
    throw new Error("ci.config.json commit branch must be non-empty.");
  }

  return { toggle, branch };
}

function scanForViolations(dir) {
  let foundViolation = false;
  const entries = fs.readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (!["node_modules", ".git"].includes(entry.name)) {
        foundViolation = scanForViolations(fullPath) || foundViolation;
      }
      continue;
    }

    if (!entry.name.endsWith(".sh")) continue;

    const content = fs.readFileSync(fullPath, "utf8");
    if (content.includes("EXIT_RUNTIME") || content.includes("--exit-runtime")) {
      console.error(`VIOLATION: ${fullPath}`);
      foundViolation = true;
    }
  }

  return foundViolation;
}

function resolveRefName() {
  return (
    process.env.GITHUB_HEAD_REF ||
    process.env.GITHUB_REF_NAME ||
    ""
  ).trim();
}

function run() {
  const configPath = path.join(process.cwd(), "ci.config.json");
  if (!fs.existsSync(configPath)) {
    console.error("Error: ci.config.json not found.");
    process.exit(1);
  }

  const config = JSON.parse(fs.readFileSync(configPath, "utf8"));
  const { toggle, branch } = parseCommitConfig(config);
  const currentRef = resolveRefName();

  setOutput("status", "skipped");
  setOutput("commit_enabled", toggle);
  setOutput("commit_branch", branch);
  setOutput("current_branch", currentRef);

  if (config.ci !== true) {
    console.log(`CI disabled. Value: ${config.ci}`);
    return;
  }

  if (toggle !== "on") {
    console.log(`Commit gate disabled. commit[0]=${toggle}`);
    return;
  }

  if (currentRef !== branch) {
    console.log(`Branch gate blocked. current=${currentRef} expected=${branch}`);
    return;
  }

  console.log(`CI enabled on branch ${currentRef}. Scanning repository...`);

  if (scanForViolations(process.cwd())) {
    process.exit(451);
  }

  setOutput("status", "ready");
  console.log("Signal sent: READY");
}

run();

File: .github/scripts/ci_check.py
import json
import os
import sys
import glob

def run_ci():
    # 1. FIND THE CONFIG (Search entire repo)
    config_matches = glob.glob('**/ci.config.json', recursive=True)
    if not config_matches:
        print("❌ Error: ci.config.json not found anywhere!")
        sys.exit(1)
    
    config_path = config_matches[0]
    print(f"🔍 Using config: {config_path}")

    # 2. LOAD AND CHECK THE TOGGLE
    with open(config_path, 'r') as f:
        try:
            data = json.load(f)
        except:
            print("❌ Error: Invalid JSON!")
            sys.exit(1)

    # Use .get() and compare strictly to True
    # This handles "false", null, or missing keys correctly
    is_enabled = data.get('ci')
    
    if is_enabled is not True:
        print(f"⏭️  CI is DISABLED (Value is: {is_enabled}). Skipping scan.")
        return

    # 3. RUN THE SCAN
    print("🚀 CI is ENABLED. Scanning...")
    found_violation = False
    for sh_file in glob.iglob('**/*.sh', recursive=True):
        with open(sh_file, 'r', errors='ignore') as f:
            if "EXIT_RUNTIME" in f.read():
                print(f"❌ VIOLATION: {sh_file}")
                found_violation = True

    if found_violation:
        sys.exit(451)
    print("✔ All clear.")

if __name__ == "__main__":
    run_ci()
